### PR TITLE
APEX-267 APEX BLS Precompile

### PR DESCRIPTION
**Pull request by igorcrevar**
# Description

Please provide a detailed description of what was done in this PR

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Checklist

- [ ] I have assigned this PR to myself
- [ ] I have added at least 1 reviewer
- [ ] I have added the relevant labels
- [ ] I have updated the official documentation
- [ ] I have added sufficient documentation in code

## Testing

- [ ] I have tested this code with the official test suite
- [ ] I have tested this code manually

### Manual tests

Please complete this section if you ran manual tests for this functionality, otherwise delete it

# Documentation update

Please link the documentation update PR in this section if it's present, otherwise delete it

# Additional comments

Please post additional comments in this section if you have them, otherwise delete it

### CODE REVIEW

File: state/runtime/precompiled/apex_bls_signs_verification.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+	apexBLSInputDataSingleABIType = abi.MustNewType("tuple(bytes32, bytes, uint256[4])")
+)
+
+// apexBLSSignatureVerification verifies the given aggregated signatures using the default BLS utils functions.
+// apexBLSSignatureVerification returns ABI encoded boolean value depends on validness of the given signatures.
+type apexBLSSignatureVerification struct {
+	domain []byte
+}
+
+// gas returns the gas required to execute the pre-compiled contract
```
Comment (sasa-ethernal) [inline state/runtime/precompiled/apex_bls_signs_verification.go:37]: "gas" at the beginning is unnecessary

File: state/runtime/precompiled/apex_bls_signs_verification.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+func (c *apexBLSSignatureVerification) gas(input []byte, _ *chain.ForksInTime) uint64 {
+	return 50000
+}
+
+// Run runs the precompiled contract with the given input.
+// Input must be ABI encoded:
+// - if first byte is 0 then other bytes are decoded as tuple(bytes32, bytes, uint256[4])
+// - otherwise other input bytes are decoded as tuple(bytes32, bytes, uint256[4][], bytes, uint8)
```
Comment (sasa-ethernal) [inline state/runtime/precompiled/apex_bls_signs_verification.go]: comment doesn't match with type, uint8 is not used
apexBLSInputDataMultiABIType = abi.MustNewType("tuple(bytes32, bytes, uint256[4][], bytes)"

File: state/runtime/precompiled/apex_bls_signs_verification_test.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+		return append([]byte{apexBLSSingleTypeByte}, encoded...)
+	}
+
+	encodeMulti := func(hash []byte, signature []byte, publicKeys [][4]*big.Int, bmp bitmap.Bitmap) []byte {
+		encoded, err := abi.Encode([]interface{}{hash, signature, publicKeys, bmp},
+			apexBLSInputDataMultiABIType)
+		require.NoError(t, err)
+
+		return append([]byte{0x76}, encoded...)
```
Comment (sasa-ethernal) [inline state/runtime/precompiled/apex_bls_signs_verification_test.go]: define apexBLSMultiTypeByte type for 0x76, just like apexBLSSingleTypeByte

**Review by sasa-ethernal — CHANGES_REQUESTED** [review]

File: state/runtime/precompiled/apex_bls_signs_verification.go (hunk 1 above)
Comment (igorcrevar) [reply state/runtime/precompiled/apex_bls_signs_verification.go:37]: that is the name of the function

File: state/runtime/precompiled/apex_bls_signs_verification_test.go (hunk 3 above)
Comment (igorcrevar) [reply state/runtime/precompiled/apex_bls_signs_verification_test.go]: https://github.com/Ethernal-Tech/blade-apex-bridge/pull/39/commits/c5f80ee1740e53b95f5d3b8c8d57a52499f44caa

File: state/runtime/precompiled/apex_bls_signs_verification.go (hunk 2 above)
Comment (igorcrevar) [reply state/runtime/precompiled/apex_bls_signs_verification.go]: https://github.com/Ethernal-Tech/blade-apex-bridge/pull/39/commits/c5f80ee1740e53b95f5d3b8c8d57a52499f44caa

**Review by sasa-ethernal — APPROVED** [review]

**Pull request merged**